### ISSUE TITLE
Path E: Generalized Teacher Forcing for autoregressive inference stabilization

### ISSUE THREAD

**Issue by Polichinel** [views-platform/views-hydranet#37]
## Sequence Position

> **This is Step 2 of 6** in the zero-inflation instability remediation sequence: **A → E → D/B → C/F**
>
> **Depends on:** #36 (Path A: Tobit loss) — must pass isolation test first.
> Path A fixes gradient starvation during training. This issue fixes error compounding during inference. They solve different problems and are complementary.

---

### The exposure bias problem

The model is trained with **teacher forcing**: at each timestep, the true historical input is provided. During inference, the model's own predictions are fed back as input (`hydranet_inference.py:289`: `t0_autoreg = t1_pred.detach()`). The model has never seen its own predictions as input during training.

This train/inference distribution mismatch causes **error compounding**: small prediction errors at step t become part of the input at step t+1, biasing the prediction further, which becomes input at step t+2, and so on for 36 steps. Even if Path A (Tobit loss) perfectly solves zero-cell gradient starvation, this compounding remains.

### Evidence of the problem

The sweep failure mode shows this clearly: all 6 sweep runs diverge during autoregressive inference, not during training. Training loss (even with hurdle) eventually decreases. The explosion happens only when predictions feed back — the model encounters out-of-distribution inputs and produces increasingly extreme outputs.

The drift detection mechanism in `hydranet_inference.py:295-307` (magnitude guards at 100, 500) catches gross divergence but doesn't prevent the gradual drift that precedes it.

### Why fixing the loss alone isn't sufficient

Path A (Tobit) stabilizes zero-cell predictions, which removes the dominant source of divergence. But prediction errors in *active* cells still compound. Over 36 steps, a 5% per-step error in conflict-active regions can grow substantially. The CRPS at step 36 will be worse than at step 1 — this is the inherent cost of autoregression, reducible but not eliminable.

---

### Option 1: Scheduled Sampling (Bengio et al. 2015) — RECOMMENDED START

**Core idea:** During training, gradually replace ground-truth inputs with model predictions using a curriculum schedule.

```
At training step i:
    With probability ε_i: use model's own prediction ẑ_{t-1} as input
    With probability 1-ε_i: use ground truth z_{t-1} as input

    ε_i increases from 0 → ε_max over training
```

**Why it works:** The model learns to handle its own imperfect predictions as input. By the end of training, a significant fraction of inputs are self-generated, closing the train/inference distribution gap.

**Schedule options:**
- Linear: `ε_i = min(ε_max, i / warmup_steps)`
- Inverse sigmoid: `ε_i = k / (k + exp(i / k))` (recommended by Bengio et al.)
- Exponential: `ε_i = 1 - k^i`

**Implementation sketch:**

```python
# In training_engine.py, during the forward loop over timesteps:
if scheduled_sampling_enabled and t > 0:
    epsilon = get_sampling_rate(current_lesson, total_lessons, schedule='inverse_sigmoid')
    if random.random() < epsilon:
        # Use model's own prediction as input (detached)
        t0_input = t1_pred.detach()
    else:
        # Use ground truth
        t0_input = ground_truth_t
```

**Config:**
```python
'teacher_forcing': 'scheduled_sampling',
'ss_schedule': 'inverse_sigmoid',
'ss_warmup_lessons': 5,     # pure teacher forcing for first 5 lessons
'ss_epsilon_max': 0.5,       # never go fully free-running during training
```

**Source:**
- Bengio et al. (2015): "Scheduled Sampling for Sequence Prediction with Recurrent Neural Networks," NeurIPS 2015. Introduces the curriculum approach to bridging the train/inference gap for recurrent models. Shows consistent improvements on image captioning, constituency parsing, and speech recognition.
  [NeurIPS proceedings](https://papers.nips.cc/paper/2015/hash/e995f98d56967d946471af29d7bf99f1-Abstract.html)

---

### Option 2: Generalized Teacher Forcing — GTF (Hess et al. 2023) — MORE PRINCIPLED

**Core idea:** Instead of a fixed schedule, adaptively mix ground truth and predictions based on the local dynamics of the system. The mixing coefficient α is determined by the Jacobian product norm κ — a measure of how chaotic the local dynamics are.

```
z̃_t = (1 - α) · z_t + α · ẑ_t

α = max(0, 1 - 1/κ)

κ = ‖∏_{s=1}^{t} J_s‖  (product of per-step Jacobians)
```

**Why it's better than scheduled sampling:**

- **Adaptive:** When dynamics are stable (κ small), α ≈ 0 — the model trains on ground truth. When dynamics are chaotic (κ large), α → 1 — the model trains on its own predictions. This is *exactly* the regime-dependent behavior we want.
- **Provably bounded:** Hess et al. prove that GTF bounds the gradient norm even for chaotic dynamics (Lyapunov exponent > 0). This is a theoretical guarantee that scheduled sampling lacks.
- **No schedule tuning:** Scheduled sampling requires choosing a schedule function, warmup period, and ε_max. GTF has no tuning parameters — α is computed from the dynamics.

**Implementation complexity:**

The Jacobian product norm κ is expensive to compute exactly for a UNet. Practical approaches:
- **Finite difference approximation:** Perturb input by small δ, measure output change. Ratio ≈ local Jacobian norm.
- **Power iteration:** Estimate the largest singular value of the Jacobian via a few iterations.
- **Exponential moving average:** Track κ as a running estimate across training steps.

```python
# GTF implementation sketch
if gtf_enabled and t > 0:
    # Estimate Jacobian norm via finite differences
    delta = torch.randn_like(z_prev) * 1e-3
    out_base = model(z_prev, h_prev).reg
    out_perturbed = model(z_prev + delta, h_prev).reg
    local_jacobian_norm = (out_perturbed - out_base).norm() / delta.norm()
    
    # Accumulate product norm (log-space for stability)
    log_kappa += torch.log(local_jacobian_norm)
    kappa = torch.exp(log_kappa)
    
    alpha = max(0.0, 1.0 - 1.0 / kappa.item())
    z_input = (1 - alpha) * z_true_t + alpha * z_pred_t.detach()
else:
    z_input = z_true_t
```

**Config:**
```python
'teacher_forcing': 'gtf',
'gtf_warmup_lessons': 3,    # pure teacher forcing for first 3 lessons
'gtf_jacobian_method': 'finite_diff',  # or 'power_iteration'
```

**Source:**
- Hess et al. (2023): "Generalized Teacher Forcing for Learning Chaotic Dynamics," ICML 2023. Proves that adaptive α = max(0, 1-1/κ) bounds gradients for chaotic systems. Validated on Lorenz-63, Lorenz-96, and Kuramoto-Sivashinsky (turbulence). Outperforms scheduled sampling and standard teacher forcing on all benchmarks.
  [ICML proceedings](https://proceedings.mlr.press/v202/hess23a.html)

---

## Recommendation

**Start with Scheduled Sampling** (Option 1). It's simpler to implement, has no computational overhead from Jacobian estimation, and addresses the core problem. If scheduled sampling proves insufficient (CRPS still degrades significantly at long horizons), upgrade to GTF (Option 2).

The Jacobian estimation in GTF adds one additional forward pass per training timestep, which roughly doubles training time. For a model that already trains for ~20 curriculum lessons with 3 windows each, this is a significant cost. Justify it only if scheduled sampling leaves measurable long-horizon quality on the table.

---

### Phase 1: Infrastructure

1. Add `teacher_forcing` config parameter to `config_initializer.py` (valid values: `'none'`, `'scheduled_sampling'`, `'gtf'`)
2. Add schedule parameters: `ss_schedule`, `ss_warmup_lessons`, `ss_epsilon_max`
3. Implement `ScheduledSamplingMixer` class in a new `views_hydranet/train/teacher_forcing.py`

### Phase 2: Training loop integration

Modify the training forward pass in `training_engine.py` to support mixed inputs. Currently the training loop processes a single window of ground truth. The change:

```
Before: z_t = ground_truth[t]        (always)
After:  z_t = mixer(ground_truth[t], pred[t-1], epsilon)  (curriculum-dependent)
```

This requires the training loop to run autoregressively during training (at least partially), which is a structural change from the current implementation.

### Phase 3: Validation

1. **Short-horizon test:** CRPS at steps 1-6 should not degrade vs Path A alone
2. **Long-horizon test:** CRPS at steps 30-36 should improve vs Path A alone
3. **Convergence:** Training should still converge within 20 lessons
4. **Drift detection:** Fewer drift warnings during inference

### Phase 4: GTF upgrade (conditional)

If scheduled sampling leaves long-horizon quality gaps, implement GTF:
1. Add Jacobian estimation method
2. Replace fixed schedule with adaptive α
3. Validate computational overhead is acceptable

---

## Interaction with Other Paths

- **Path A (Tobit):** Prerequisite. The loss function determines what the model learns; teacher forcing determines what inputs it sees during learning. Both are needed.
- **Path B (Heckman):** Independent. Heckman couples the classification and regression heads — orthogonal to how training inputs are constructed.
- **Path D (Tail-aware):** Independent. Tail calibration is about the loss function for extreme values — orthogonal to training input mixing.
- **Path F (Direct MH):** Mutually exclusive. If we switch to direct multi-horizon forecasting, there is no autoregressive loop and teacher forcing is irrelevant.

---

## Acceptance Criteria

- [ ] `ScheduledSamplingMixer` implemented with configurable schedule
- [ ] Training loop supports mixed ground-truth / prediction inputs
- [ ] Config schema accepts `teacher_forcing` parameter
- [ ] Scheduled sampling validated: short-horizon CRPS preserved, long-horizon CRPS improved
- [ ] All existing tests pass
- [ ] (Optional) GTF variant implemented and benchmarked against scheduled sampling
- [ ] ADR documenting the teacher forcing approach and schedule choice

**Comment by Polichinel** [comment]
## Path E — Experimentally Validated (Gate 2 PASS)

### Scheduled Sampling Sweep Results (2026-06-01)

Linear schedule, warmup=10 lessons, per-target sigma {sb: 1.0, ns: 0.75, os: 0.5}, 80 lessons.

**Step-wise metrics:**
| eps_max | sb CRPS | sb MCR | ns CRPS | ns MCR | os MCR |
|---------|---------|--------|---------|--------|--------|
| 0.00 (control) | 0.265 | 1.92 | 0.037 | 0.36 | 0.20 |
| 0.25 | 0.200 | **1.01** | 0.036 | 0.33 | 0.11 |
| **0.50** | **0.152** | 0.37 | **0.031** | 0.05 | 0.23 |
| 0.75 | 0.146 | 0.27 | 0.031 | 0.002 | 0.15 |

**Exposure bias gap (sb MCR: month vs step):**
- Control: 0.33 vs 1.92 = gap of **1.60**
- eps=0.50: 0.35 vs 0.37 = gap of **0.02** (eliminated)

### Gate 2 PASSES

The scheduled sampling sweep demonstrates:
- Exposure bias gap: 1.60 → 0.02 (99% reduction)
- Step-wise sb CRPS: 0.265 → 0.152 (43% improvement)
- No escalation to GTF (Hess et al. 2023) needed

**Recommended production config:** `ss_schedule='linear', ss_warmup_lessons=10, ss_epsilon_max=0.5`

### Literature
- Bengio, S. et al. (2015). "Scheduled Sampling for Sequence Prediction with Recurrent Neural Networks." NeurIPS 2015.
- Hess, F. et al. (2023). "Generalized Teacher Forcing for Learning Chaotic Dynamics." ICML 2023. (escalation path — not needed)

### Implementation
- PR #50 merged (ADR-056)
- `ScheduledSamplingMixer` with 3 schedules (linear, inverse_sigmoid, exponential)
- 24 tests, 703 total passing
